orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -317,17 +321,54 @@ IMPLEMENT_GRADIENT_BUILDER(GetMatMulGradient) {
       HandleBroadcastingDynamic(pre_reduce_grad_0, A, a_shape, GI(0), a_axes, result);
     }
     if (IsGradientRequiredForSrcNodeInput(1)) {
-      ArgDef pre_reduce_grad_1 = IA("PreReduceGrad1");
-      result.push_back(
-          NodeDef(OpDef{"TransposeScaleMatMul", kMSDomain, 1},
-                  {A, GO(0)},
-                  {pre_reduce_grad_1},
-                  {{"transA", MakeAttribute("transA", int64_t(1))}}));
-
-      b_axes = IA("ReduceAxes_" + B.name + "_for_" + B.name);
-      ia_shape = IA("Shape_" + pre_reduce_grad_1.name);
-      ComputeBroadcastBackwardAxesDynamic(pre_reduce_grad_1, B, ia_shape, b_shape, nullptr, &b_axes, result);
-      HandleBroadcastingDynamic(pre_reduce_grad_1, B, b_shape, GI(1), b_axes, result);
+      if (B_has_shape && B_shape.size() == 2) {
+        // for case: A[M1, M2, ... , K], B[K, N], Y[M1, M2, ..., N]
+        NodeDef zero_int64_const_node = ConstantScalarNode(int64_t{0}, {1}, Name("zero_int64"));
+        NodeDef one_const_node = ConstantScalarNode(int64_t{1}, {1}, Name("one"));
+        NodeDef neg_one_const_node = ConstantScalarNode(int64_t{-1}, {1}, Name("neg_one"));
+        NodeDef zero_float_const_node = ConstantScalarNode(float{0.0f}, {1}, Name("zero_float"));
+
+        ArgDef ZERO_I = zero_int64_const_node.output_args[0];
+        ArgDef ONE = one_const_node.output_args[0];
+        ArgDef NEG_ONE = neg_one_const_node.output_args[0];
+        ArgDef ZERO_F = zero_float_const_node.output_args[0];
+
+        std::vector<NodeDef> dB_subgraph{
+            zero_int64_const_node,
+            one_const_node,
+            neg_one_const_node,
+            zero_float_const_node,
+
+            NodeDef("Shape", {B}, {IA("B_shape")}),
+
+            // reshape A to 2D [M, K]
+            NodeDef("Gather", {IA("B_shape"), ZERO_I}, {IA("K_dim")}, {MakeAttribute("axis", int64_t(0))}),
+            NodeDef("Concat", {NEG_ONE, IA("K_dim")}, {IA("A_target_shape")}, {MakeAttribute("axis", int64_t(0))}),
+            NodeDef("Reshape", {A, IA("A_target_shape")}, {IA("A_reshape_2d")}),
```

Comment:
Can you just reshape it to [-1 0] ?

Reply:
no... reshaping [500, 4, 1024] with [-1, 0] will end in [500x1024, 4]. 0 is indicating following left aligned shape. 
We need target shape [500x4, 1024].

Reply:
Squeeze(Reshape(x, [1,-1,0]), 0) should do.

---
In reply to: [475955739](https://github.com/microsoft/onnxruntime/pull/4899#discussion_r475955739) [](ancestors = 475955739)

Reply:
x could be arbitrary dimension. eg. for 4D x, you would need [1,1,-1,0] instead.

Reply:
Flatten(x, axis=-1)?
[https://github.com/onnx/onnx/blob/master/docs/Operators.md#Flatten](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Flatten)

---
In reply to: [475967015](https://github.com/microsoft/onnxruntime/pull/4899#discussion_r475967015) [](ancestors = 475967015)

Reply:
Thanks a lot!!! Flatten will do the trick here!